Add FAQ and Troubleshooting docs for Machine ID (#14871)
gravitational/teleport: docs/pages/machine-id/troubleshooting.mdx

---
title: Machine ID Troubleshooting Guide
description: Troubleshooting common issues with Machine ID
---

> ## A bot failed to renew a certificate due to a "generation mismatch", what does this mean?

Reviewer: ```suggestion
## A bot failed to renew a certificate due to a "generation mismatch"
```
To be consistent with the other headings

### Symptoms

The bot will log an error like this:

```text
ERROR: renewable cert generation mismatch: stored=3, presented=2
```

Subsequent connection attempts by the bot may see errors like the following:
```text
ERROR: failed direct dial to auth server: auth API: access denied [00]
"\tauth API: access denied [00], failed dial to auth server through reverse tunnel: Get \"https://teleport.cluster.local/v2/configuration/name\": Get \"https://example.com:3025/webapi/find\": x509: cannot validate certificate for example.com because it doesn't contain any IP SANs"
"\tGet \"https://teleport.cluster.local/v2/configuration/name\": Get \"https://example.com:3025/webapi/find\": x509: cannot validate certificate for example.com because it doesn't contain any IP SANs"
```

> In particular, note the message `auth API: access denied`. The Teleport auth

Reviewer: ```suggestion
In particular, note the message `auth API: access denied`. The Teleport Auth
```

> server will provide some additional context:

Reviewer: ```suggestion
Service will provide some additional context:
```

```text

> [AUTH]      WARN lock targeting User:"bot-example" is in force: The bot user "bot-example" has been locked due to a certificate generation mismatch, possibly indicating a stolen certificate. auth/apiserver.go:224

Reviewer: Is this a log on the Machine ID host? Otherwise, we might want to use a `ScopedBlock` to hide the message about this log for Cloud users.

Author: Ah, good catch, I'll add a `ScopedBlock`

```

### Explanation

Machine ID (with token-based joining) uses a certificate generation counter to
detect potentially stolen renewable certificates. Each time a bot fetches a new

> renewable certificate, the counter is incremented and stored on the Teleport

Reviewer: > the counter is incremented and stored on the Teleport
Auth Server and embedded in the certificate itself.

Suggesting something like this to tighten this up and clarify the architecture:

```markdown
the Auth Service increments the counter, stores it on the
backend, and embeds the counter in the certificate.
```

Author: That's much better, thanks! I think I'll add a small extra detail to mention that we embed "[a copy] of the counter in the certificate" to imply the cert isn't the source of truth (one of these days I need to write up an architecture doc).

Auth Server and embedded in the certificate itself.

If the counter embedded in your bot certificate doesn't match the counter
stored in Teleport's Auth Server, the renewal will fail and the bot user will
be automatically [locked](../access-controls/guides/locking.mdx). 

Renewable certificates are exclusively stored in the bot's internal data
directory, by default `/var/lib/teleport/bot`. It's possible to trigger this by

> accident if multiple bots are started using same internal data directory, or if

Reviewer: ```suggestion
accident if multiple bots are started using the same internal data directory, or if
```

this internal data is otherwise being shared between multiple bot instances.

Additionally, if a bot fails to save its freshly renewed certificates (for
example, due to a filesystem error) and crashes, it will attempt a renewal
with old certificates and trigger a lock.

### Resolution

Before unlocking the bot, try to determine if either of the two scenarios
described above apply. If the certificates were stolen, there may be 
underlying security concerns that need to be addressed.

Otherwise, first ensure only one bot instance is using the internal data
directory. Multiple bots can be run on a single system, but separate data
directories must be configured for each.

> Additionally, ensure the internal data is not being copied to any other nodes.

Reviewer: When would this happen?

Author: I've expended on this somewhat and clarified what a destination directory is.

If you'd like to share certificates between nodes, only copy content from
destination directories.

> Once the underlying cause has been addressed, two steps required to reset a

Reviewer: ```suggestion
Once you have addressed the underlying cause, follow these steps to reset a
```
There's a lot of passive voice on this page, so I've suggested making this sentence more active.

ERRO [TBOT]      Failed to resolve tunnel address Get "https://auth.example.com:3025/webapi/find": x509: cannot validate certificate for auth.example.com because it doesn't contain any IP SANs reversetunnel/transport.go:90
ERROR: failed direct dial to auth server: Get "https://teleport.cluster.local/v2/configuration/name": remote error: tls: bad certificate
"\tGet \"https://teleport.cluster.local/v2/configuration/name\": remote error: tls: bad certificate, failed dial to auth server through reverse tunnel: Get \"https://teleport.cluster.local/v2/configuration/name\": Get \"https://auth.example.com:3025/webapi/find\": x509: cannot validate certificate for auth.example.com because it doesn't contain any IP SANs"
"\tGet \"https://teleport.cluster.local/v2/configuration/name\": Get \"https://auth.example.com:3025/webapi/find\": x509: cannot validate certificate for auth.example.com because it doesn't contain any IP SANs"
```

In particular, note the log line: "Identity has expired. The renewal is likely to
fail."

### Explanation

> Token-joined bots are unable to reauthenticate to the Teleport server once their

Reviewer: Which Teleport server does this refer to?

Author: I'll clarify this to be the Teleport Auth Service. My only qualm (?) is that we'll also happily connect to the Proxy Service as well, but I don't think that'll really be a source of confusion..

certificates have expired. Tokens in token-based joining (as opposed to AWS IAM

> joining) are one-time-use, so when the bot's internal certificates expire, it

Reviewer: ```suggestion
joining) can only be used once, so when the bot's internal certificates expire, it
```

will not be able to connect.

> When a bot's identity expires, certain server-side parameters like the

Reviewer: I think using the active voice here would help clarify where this is taking place, maybe something like:

"When a bot's identity expires, the Auth Service must reset the certificate generation counter and other parameters".

Author: This is unfortunately a manual process so I'm not sure it's appropriate to say the auth service is really performing any action here. I've tried to clarify things but might just be digging myself into a passive voice hole.

certificate generation counter must be reset. The simplest way to accomplish
this is by removing and recreating the bot, which purges all server-side data
and issues a new joining token.

### Resolution

Remove and recreate the bot, replacing the name and role list as desired:

```code
$ sudo tctl bots rm example
$ sudo tctl bots add example --roles=access
```

> Copy the resulting join token into the existing bot config - either the

Reviewer: ```suggestion
Copy the resulting join token into the existing bot config—either the
```

> `--token` CLI flag or the `onboarding.token` parameter in `tbot.yaml` - and

Reviewer: ```suggestion
`--token` CLI flag or the `onboarding.token` parameter in `tbot.yaml`—and
```
I think em dashes look better here

restart the bot. It will detect the new token and rejoin the cluster as normal.

## SSH connections fail with `ssh: handshake failed: ssh: unable to authenticate`

### Symptoms

When attempting to connect to a node via SSH, connections fail with an error
like the following:

```code
$ ssh -F /opt/machine-id/ssh_config bob@node.example.com                                                                                                                           timothyb89/v10/machine-id-app-access 
ERROR: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain

ERROR: unable to execute tsh
executing `tsh proxy`
exit status 1

kex_exchange_identification: Connection closed by remote host
Connection closed by UNKNOWN port 65535
```

In particular, note the `ssh: unable to authenticate` message.

### Explanation

This can occur when attempting to log into the node as a user not listed as a
principal on the SSH certificate.

> This can be verified by viewing the `tbot` logs and looking for the log message

Reviewer: ```suggestion
You can verify this by viewing the `tbot` logs and looking for the log message
```

when impersonated certificates for the matching destination were renewed.

In the following example, the only principal listed for the identity in
`/opt/machine-id` is `alice` (via the `access` role):
```text
INFO [TBOT]      Successfully renewed impersonated certificates for directory /opt/machine-id, valid: after=2022-07-21T21:49:26Z, before=2022-07-21T22:50:26Z, duration=1h1m0s | kind=tls, renewable=false, disallow-reissue=true, roles=[access], principals=[alice -teleport-internal-join], generation=0 tbot/renew.go:630
```

> ... however, the SSH command attempted to login as `bob`.

Reviewer: ```suggestion
However, the SSH command attempted to log in as `bob`.
```

### Resolution

> Ensure the bot identity is allowed to login as the requested user by taking any

Reviewer: ```suggestion
Ensure the bot identity is allowed to log in as the requested user by taking any
```
"login" is a noun

of the following actions:

>  1. Changing the SSH command to login as an allowed user

Reviewer: ```suggestion
 1. Changing the SSH command to log in as an allowed user
```

 2. Modifying the `access` role to allow the `alice` principal

>  3. Adding an additional role granting login via the `bob` principal

Reviewer: ```suggestion
 3. Adding a role granting login via the `bob` principal
```
No need for "adding an additional"

Note that if roles are added or modified, the certificates will need to be
renewed for the changes to take effect. The bot will renew certificates on its
own after the renewal interval (by default, 20 minutes), but you can trigger a

> renewal immediately by either restarting the `tbot` process or sending it a

Reviewer: What does "restarting the `tbot` process" mean here? `systemctl restart tbot`?

Author: I've added an additional `systemctl` example to the code block to hopefully clarify this.

reload signal:

```code
$ pkill -sigusr1 tbot
```

## Database requests fail with `database "example" not found`, but the database exists

### Symptoms

> When requesting [database access](../database-access/introduction.mdx)

Reviewer: ```suggestion
When requesting [Database Access](../database-access/introduction.mdx)
```

certificates, the certificate request fails with an error like the following:

```text
ERROR: Failed to generate impersonated certs for directory /opt/machine-id: database "example" not found
database "example" not found
```

However, the database exists and can be seen by regular users via `tsh`:

```code
$ tsh db ls
Name       Description Allowed Users Labels  Connect 
---------- ----------- ------------- ------- ------- 
example                [alice]       env=dev
```

### Explanation

Unlike regular Teleport users, Machine ID bot users are granted only minimal
Teleport [RBAC permissions](../access-controls/reference.mdx) and are not
allowed to view or list databases by default unless granted permission via one

> or more roles. Users should grant these permissions to the bot via a role at

Reviewer: I think we should remove the last sentence of this paragraph since the "Resolution" section provides this information with much more detail.

first setup.

### Resolution

Per the [Machine ID Database Access Guide](./guides/databases.mdx), ensure at
least one role providing database permissions has been granted to the
destination listed in the error.

> <Notice type="note">

Reviewer: - This Notice seems to interrupt the paragraphs above/below it.

- What is a destination in this context?

Author: I'll move the notice down a bit, and I'll try to disambiguate destinations a bit. Destinations come up a lot and I imagine a  more appropriate place to define that would be a dedicated architecture page.

By default, destinations are granted all roles provided to the bot via
`tctl bots add --roles=...`, but it's possible to grant only a subset of these
roles using the `roles: ...` parameter in `tbot.yaml`.

If permissions are unexpectedly missing, ensure `tbot.yaml` requests the
database role, either by relying on default behavior or adding the role to the
`roles: ...` list.
</Notice>

> An example role might look like the following:

Reviewer: What is it about this example that we are illustrating here?

Author: I've clarified this a bit.
